jupyter-js-widgets/src/widget.ts

```diff
@@ -399,23 +406,82 @@ class WidgetModel extends Backbone.Model {
+            function seperate_buffers(obj, path) {
+                if(visited_set.indexOf(obj) != -1) { // we already visited this object
+                    return obj;
+                }
+                visited_set.push(obj);
+                if (_.isArray(obj)) {
+                    var is_cloned = false;
+                    for (var i = 0; i < obj.length; i++) {
+                        var value = obj[i];
+                        if(value) {
+                            if (value.buffer instanceof ArrayBuffer || value instanceof ArrayBuffer) {
+                                if(!is_cloned) {
+                                    obj = _.map(obj, _.identity);
+                                    is_cloned = true;
+                                }
+                                buffers.push(value);
+                                buffer_paths.push(path.concat([i]));
+                                // easier to just keep the array, but clear the entry, otherwise we have to think
+                                // about array length
+                                obj[i] = null;
+                            }
+                            else {
+                                var new_value  = seperate_buffers(value, path.concat([i]));
+                                if((new_value != value) && !is_cloned) { // only clone when we have to
+                                    obj = _.map(obj, _.identity);
+                                    is_cloned = true;
+                                    obj[i] = new_value;
+                                }
+                            }
+                        }
                     }
                 }
+                else if(_.isObject(obj)) {
+                    for (var key in obj) {
+                        var is_cloned = false;
+                        if (!obj.hasOwnProperty || obj.hasOwnProperty(key)) {
+                            var value = obj[key];
+                            if(value) {
+                                if (value.buffer instanceof ArrayBuffer || value instanceof ArrayBuffer) {
+                                    if(!is_cloned) {
+                                        obj = _.mapObject(obj, _.identity); // clone only once for performance
+                                        is_cloned = true;
+                                    }
+                                    buffers.push(value);
+                                    buffer_paths.push(path.concat([key]));
+                                    delete obj[key]; // for objects/dicts we just delete them
+                                }
+                                else {
+                                    var new_value  = seperate_buffers(value, path.concat([key]));
+                                    if((new_value != value) && !is_cloned) { // only clone when we have to
+                                        obj = _.mapObject(obj, _.identity);
```

Comment:
I almost wrote that, but in this case, we know we're not a primitive or an array, right? `_.clone` degenerates to `_.extend` after it checks for other types.